### ISSUE TITLE
No automatic refresh handling?

### ISSUE THREAD

**Issue by djMax** [Dwolla/dwolla-v2-node#16]
This is a fairly thin library, which I suppose in general is a good thing, but I'd expect it to automatically manage token refresh... It doesn't look like it does, but any plans?

**Comment by sausman** [comment]
At the moment we don't have any plans to automatically manage token refresh, but the library does allow you to configure an `onGrant` callback that you can use to store tokens whenever one is granted. Since tokens need to be stored somewhere it's probably out of scope for this library to manage them entirely, but we're definitely open to ideas. Going forward I think we're going to focus our efforts on eliminating the need to manage tokens altogether though (while still maintaining our security requirements).

We recently made a change that allows [app tokens](https://github.com/Dwolla/dwolla-v2-node#application-tokens) to access white label customer related endpoints:

> If an application has the ManageCustomers scope enabled, it can also be used to access the API for White Label Customer related endpoints. Keep in mind, the application must belong to same Dwolla account that will be used when creating and managing White Label Customers in the API.

These tokens still expire so you'll need to manage them somehow in your application, but instead of refreshing them you can use your client credentials to generate another one (instead of using a refresh_token).

**Comment by djMax** [comment]
I think what could be a useful middle ground is some event or opportunity to supply a new token and retry a request when a token expiration is detected. i.e. I'd rather not muddy up client code with token refresh concerns, but I get that it's not the APIs job to manage the tokens themselves. I ended up doing this:

```
  let token = await dwolla.auth.client();
  let inFlightRequest = null;

  async function shouldRetry(result) {
    if (result.status === 401) {
      if (!inFlightRequest) {
        inFlightRequest = dwolla.auth.client();
        winston.info('Refreshing Dwolla token');
      }
      const originalRequest = inFlightRequest;
      token = await inFlightRequest;
      if (originalRequest === inFlightRequest) {
        inFlightRequest = null;
        winston.info('Dwolla token refresh complete');
      }
      return true;
    }
    return false;
  }

  const client = {
    async post(url, request) {
      const result = await token.post(url, request);
      if (await shouldRetry(result)) {
        return await token.post(url, request);
      }
      return result;
    },
    async get(url, request) {
      const result = await token.get(url, request);
      if (await shouldRetry(result)) {
        return await token.get(url, request);
      }
      return result;
    },
  };
```

Which I still need to fix up once I figure out how to properly detect token expiration. What is the right error/status/message to look for to now a refresh should be done? It's a bit annoying to test because you have to wait an hour, and what developer has that kind of attention span? :)

**Comment by sausman** [comment]
Haha, you should get back a 401 with the error code `ExpiredAccessToken`. If it helps here's a list of some other common error statuses/codes: https://docsv2.dwolla.com/#errors

Thanks for the detailed code examples. If we add that functionality to the library I will try to remember to update this issue letting you know.

**Comment by sausman** [comment]
Fixed in #33 (version 3.x)
